pytorch_lightning/trainer/connectors/checkpoint_connector.py

```diff
@@ -169,12 +170,16 @@ def restore_training_state(self) -> None:
         if not self._loaded_checkpoint:
             return
 
-        # restore precision plugin (scaler etc.)
-        self.trainer.precision_plugin.on_load_checkpoint(self._loaded_checkpoint)
         # restore loops and their progress
         self.restore_loops()
 
-        self.restore_optimizers_and_schedulers()
+        # restore precision plugin (scaler etc.)
+        self.trainer.precision_plugin.on_load_checkpoint(self._loaded_checkpoint)
+
+        assert self.trainer.state.fn is not None
```

Comment:
why do we need the asserts?

Reply:
mypy I guess
https://github.com/PyTorchLightning/pytorch-lightning/pull/9413#discussion_r711102946